lib/class-wp-block-templates-registry.php

```diff
@@ -0,0 +1,272 @@
+		 */
+		private $registered_block_templates = array();
+
+		/**
+		 * Container for the main instance of the class.
+		 *
+		 * @since 6.7.0
+		 * @var WP_Block_Templates_Registry|null
+		 */
+		private static $instance = null;
+
+		/**
+		 * Registers a block template.
+		 *
+		 * @since 6.7.0
+		 *
+		 * @param string|WP_Block_Template $template_name Block template name including namespace, or alternatively
+		 *                                                a complete WP_Block_Template instance. In case a WP_Block_Template
+		 *                                                is provided, the $args parameter will be ignored.
+		 * @param array                    $args          Optional. Array of block template arguments.
+		 * @return WP_Block_Template|false The registered block template on success, or false on failure.
+		 */
+		public function register( $template_name, $args = array() ) {
+
+			$template = null;
+			if ( $template_name instanceof WP_Block_Template ) {
+				$template      = $template_name;
+				$template_name = $template->name;
+			}
+
+			if ( ! is_string( $template_name ) ) {
+				_doing_it_wrong(
+					__METHOD__,
+					__( 'Block template names must be a string.', 'gutenberg' ),
+					'6.7.0'
+				);
+				return new WP_Error( 'template_name_no_string', __( 'Block template names must be a string.', 'gutenberg' ) );
+			}
+
+			if ( preg_match( '/[A-Z]+/', $template_name ) ) {
+				_doing_it_wrong(
+					__METHOD__,
+					__( 'Block template names must not contain uppercase characters.', 'gutenberg' ),
+					'6.7.0'
+				);
+				return new WP_Error( 'template_name_no_uppercase', __( 'Block template names must not contain uppercase characters.', 'gutenberg' ) );
+			}
+
+			$name_matcher = '/^[a-z0-9-]+\/\/[a-z0-9-]+$/';
+			if ( ! preg_match( $name_matcher, $template_name ) ) {
+				_doing_it_wrong(
+					__METHOD__,
+					__( 'Block template names must contain a namespace prefix. Example: my-plugin//my-custom-template', 'gutenberg' ),
+					'6.7.0'
+				);
+				return new WP_Error( 'template_no_prefix', __( 'Block template names must contain a namespace prefix. Example: my-plugin//my-custom-template', 'gutenberg' ) );
+			}
+
+			if ( $this->is_registered( $template_name ) ) {
+				_doing_it_wrong(
+					__METHOD__,
+					/* translators: %s: Template name. */
+					sprintf( __( 'Template "%s" is already registered.', 'gutenberg' ), $template_name ),
+					'6.7.0'
+				);
+				/* translators: %s: Template name. */
+				return new WP_Error( 'template_already_registered', __( 'Template "%s" is already registered.', 'gutenberg' ) );
+			}
+
+			if ( ! $template ) {
+				$theme_name = get_stylesheet();
+				$slug       = isset( $args['slug'] ) ? $args['slug'] : explode( '//', $template_name )[1];
+
+				$template                 = new WP_Block_Template();
+				$template->id             = $theme_name . '//' . $slug;
+				$template->theme          = $theme_name;
```

Comment:
I guess using the "theme" here is mandatory at the moment for the site editor right? Because ideally, plugin registered templates shouldn't be tied to a particular theme (but that probably requires a lot of work separately).

Reply:
Good question. Removing the `theme` property causes [some errors](https://github.com/WordPress/wordpress-develop/blob/606fcbd72526021b5d950809b1068f605a2bac73/src/wp-includes/block-template-utils.php#L816) when saving the template in the Site Editor. My understanding is that the concept of "theme-agnostic" templates doesn't exist, as any template needs to be attached to a theme, so when switching from one theme to another it gets reset. :thinking:

Reply:
Yes, I think we're yet to implement that, we can reassess this later.